### ISSUE TITLE
Compilation fails with 20.04.1 LTS and CMake but passes with autotools

### ISSUE THREAD

**Issue by drhenault** [libsndfile/libsndfile#665]
Many thanks to the team for making this useful lib available. Unfortunately, compilation fails with Ubuntu 20.04.1 LTS and CMake but passes with autotools. This is annoying because we are using libsndfile within another CMake project that manage external dependencies through CMake.

The main difference between the autotool and CMake builds seems to be the generation of the `src/config.h` file. I believe the most significant difference is the line with `#define HAVE_SNDIO_H`.

- With autotool, I have `#define HAVE_SNDIO_H 0`, which is incorrect (since I do have it), but that let me compile `libsndfile`
- With CMake, I have `#define HAVE_SNDIO_H 1`, which is correct, but then the compilation fails because `sndio.h` is not properly included (resulting in undefined symbols, e.g. `sio_par`, `sio_open`, `sio_initpar`, etc. in source file `sndfile-play.c`). *Workaround:* Manually changing the 1 for a 0 let me proceed with the build.

In both cases, I do have `snprintf`-related warning regarding truncation, but I think this is another story.

**Comment by drhenault** [comment]
Details:

**Comment by drhenault** [comment]
Details:
```
ninja
[7/127] Building C object CMakeFiles/sndfile-play.dir/programs/sndfile-play.c.o
FAILED: CMakeFiles/sndfile-play.dir/programs/sndfile-play.c.o 
/usr/bin/cc  -I/home/david/dev/external/libsndfile/src -Isrc -I/home/david/dev/external/libsndfile/include -Iinclude -std=gnu99 -MD -MT CMakeFiles/sndfile-play.dir/programs/sndfile-play.c.o -MF CMakeFiles/sndfile-play.dir/programs/sndfile-play.c.o.d -o CMakeFiles/sndfile-play.dir/programs/sndfile-play.c.o   -c /home/david/dev/external/libsndfile/programs/sndfile-play.c
/home/david/dev/external/libsndfile/programs/sndfile-play.c: In function ‘sndio_play’:
/home/david/dev/external/libsndfile/programs/sndfile-play.c:675:17: error: storage size of ‘par’ isn’t known
  675 |  struct sio_par par ;
      |                 ^~~
/home/david/dev/external/libsndfile/programs/sndfile-play.c:693:14: warning: implicit declaration of function ‘sio_open’; did you mean ‘sf_open’? [-Wimplicit-function-declaration]
  693 |   if ((hdl = sio_open (NULL, SIO_PLAY, 0)) == NULL)
      |              ^~~~~~~~
      |              sf_open
/home/david/dev/external/libsndfile/programs/sndfile-play.c:693:30: error: ‘SIO_PLAY’ undeclared (first use in this function)
  693 |   if ((hdl = sio_open (NULL, SIO_PLAY, 0)) == NULL)
      |                              ^~~~~~~~
/home/david/dev/external/libsndfile/programs/sndfile-play.c:693:30: note: each undeclared identifier is reported only once for each function it appears in
/home/david/dev/external/libsndfile/programs/sndfile-play.c:698:3: warning: implicit declaration of function ‘sio_initpar’ [-Wimplicit-function-declaration]
  698 |   sio_initpar (&par) ;
      |   ^~~~~~~~~~~
/home/david/dev/external/libsndfile/programs/sndfile-play.c:703:12: error: ‘SIO_LE_NATIVE’ undeclared (first use in this function)
  703 |   par.le = SIO_LE_NATIVE ;
      |            ^~~~~~~~~~~~~
/home/david/dev/external/libsndfile/programs/sndfile-play.c:705:9: warning: implicit declaration of function ‘sio_setpar’ [-Wimplicit-function-declaration]
  705 |   if (! sio_setpar (hdl, &par) || ! sio_getpar (hdl, &par))
      |         ^~~~~~~~~~
/home/david/dev/external/libsndfile/programs/sndfile-play.c:705:37: warning: implicit declaration of function ‘sio_getpar’ [-Wimplicit-function-declaration]
  705 |   if (! sio_setpar (hdl, &par) || ! sio_getpar (hdl, &par))
      |                                     ^~~~~~~~~~
/home/david/dev/external/libsndfile/programs/sndfile-play.c:710:9: warning: implicit declaration of function ‘sio_start’; did you mean ‘va_start’? [-Wimplicit-function-declaration]
  710 |   if (! sio_start (hdl))
      |         ^~~~~~~~~
      |         va_start
/home/david/dev/external/libsndfile/programs/sndfile-play.c:716:4: warning: implicit declaration of function ‘sio_write’ [-Wimplicit-function-declaration]
  716 |    sio_write (hdl, buffer, readcount * sizeof (short)) ;
      |    ^~~~~~~~~
/home/david/dev/external/libsndfile/programs/sndfile-play.c:718:3: warning: implicit declaration of function ‘sio_close’; did you mean ‘sf_close’? [-Wimplicit-function-declaration]
  718 |   sio_close (hdl) ;
      |   ^~~~~~~~~
      |   sf_close
[20/127] Linking CXX executable sndfilehandle
ninja: build stopped: subcommand failed.
```

**Comment by drhenault** [comment]
The problem is that inclusion of header is conditional on **not** being defined `__ANDROID__`,  `__linux__`, `__FreeBSD_kernel__` or `__FreeBSD__` and `HAVE_SNDIO_H`, whereas compilation only depends on `HAVE_SNDIO_H`
```
Extract from `./programs/sndfile-play.c`
#if defined (__ANDROID__)

#elif defined (__linux__) || defined (__FreeBSD_kernel__) || defined (__FreeBSD__)
        #include        <fcntl.h>
        #include        <sys/ioctl.h>
        #include        <sys/soundcard.h>

#elif HAVE_SNDIO_H
        #include <sndio.h>

#elif (defined (sun) && defined (unix))
        #include <fcntl.h>
        #include <sys/ioctl.h>
        #include <sys/audioio.h>

#elif (OS_IS_WIN32 == 1)
        #include <windows.h>
        #include <mmsystem.h>

#endif
```

**Comment by drhenault** [comment]
A potential solution is making sure the conditions for header inclusion and the compilation of `static void sndio_play (int argc, char *argv [])` matches.

**Comment by drhenault** [comment]
Not sure if you would prefer to update header inclusion or code compilation in `./programs/sndfile-play.c`, or to implement another fix.

**Comment by evpobr** [comment]
Hi @drhenault, thanks for report.

I guess the problem exists because Autotools checks Sndio for OpenBSD only.

**Comment by evpobr** [comment]
Check #670 please.

**Comment by drhenault** [comment]
Thanks #670 solved it for me!
